base/abstractset.jl

```diff
@@ -476,6 +476,61 @@ function isdisjoint(a, b)
     _isdisjoint(a, b)
 end
 
+function isdisjoint(a::AbstractRange{T}, b::AbstractRange{T}) where T
+    (isempty(a) || isempty(b)) && return true
+    fa, la = extrema(a)
+    fb, lb = extrema(b)
+    if (la < fb) | (lb < fa)
+        return true
+    else
+        return _overlapping_range_isdisjoint(a, b)
+    end
+end
+
+_overlapping_range_isdisjoint(a::AbstractRange{T}, b::AbstractRange{T}) where T = invoke(isdisjoint, Tuple{Any,Any}, a, b)
+
+function _overlapping_range_isdisjoint(a::AbstractRange{T}, b::AbstractRange{T}) where T<:Integer
+    m = gcd(step(a), step(b))
+    fa, fb = minimum(a), minimum(b)
+    if mod(fa - fb, m) == 0
+        if length(a) > length(b)
+            longrg = a
+            shortrg = b
+        else
+            longrg = b
+            shortrg = a
+        end
+        mis, mxs = extrema(shortrg)
+        mil, mxl = extrema(longrg)
+
+        # number of elements in shortrg that overlap longrg
+        overlapping = length(shortrg)
+
+        if mis < mil
+            overlapping -= cld(mil - mis, abs(step(shortrg)))
+        end
+        if mxs > mxl
+            overlapping -= cld(mxs - mxl, abs(step(shortrg)))
+        end
+
+        if overlapping ≥ abs(cld(step(longrg), m))
+            return false
+        else
+            if (mis < mil) & signbit(step(shortrg)) || (mxs > mxl) & ~signbit(step(shortrg))
+                return invoke(isdisjoint, Tuple{Any,Any},
+                    shortrg[begin:min(end-overlapping+1,end)], longrg)
+            elseif (mis < mil) & ~signbit(step(shortrg)) || (mxs > mxl) & signbit(step(shortrg))
+                return invoke(isdisjoint, Tuple{Any,Any},
+                    shortrg[max(begin+overlapping-1,begin):end], longrg)
+            else
+                return invoke(isdisjoint, Tuple{Any,Any}, shortrg, longrg)
+            end
+        end
```

Comment:
> You alos had a simpler implementation before computing the gcd of step sizes. That approach can get many common ranges in O(1), even outperforming this implementation's O(log(n)), at the cost of missing a few more O(n) cases.

I'm not sure I follow. Are you suggesting I abort the fast-path earlier:
```suggestion
        return invoke(isdisjoint, Tuple{Any,Any}, shortrg, longrg)
```

Reply:
I was thinking about something that would not use gcd because gcd is a slower operation. With the 1.8's `isdisjoint` implementation, we have these timings, for example:
```julia
julia> @benchmark isdisjoint(x, y) setup=(x=rand(Int):rand(Int):rand(Int); y=rand(Int):rand(Int):rand(Int))
BenchmarkTools.Trial: 10000 samples with 999 evaluations.
 Range (min … max):   7.482 ns … 624.133 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     15.956 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   22.314 ns ±  21.357 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▄▂▂ █▇▅▃▂▃▃▃▅▄▂▂▁▁       ▂▄▁▂▁           ▁                  ▂
  ██████████████████████▇▇▇▆███████▇▇▇▅▆▅▅▅██▇█▇▆▆▆▄▆▅▅▄▅▄▅▆▄▅ █
  7.48 ns       Histogram: log(frequency) by time        92 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark gcd(x, y) setup=(x=rand(Int); y=rand(Int))
BenchmarkTools.Trial: 10000 samples with 967 evaluations.
 Range (min … max):   70.770 ns … 542.424 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     105.013 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   108.666 ns ±  20.834 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

              ▁ ▄▇█ ▆▅▃                                          
  ▂▁▂▁▁▂▂▂▁▃▆██▃███▃███▄█▇▅▄▃▄▃▃▃▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂▂ ▃
  70.8 ns          Histogram: frequency by time          186 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
The input distribution is likely not representative for most use cases, but it still makes a case for not using gcd unless necessary.

Perhaps
```julia
function _overlapping_range_isdisjoint(a::AbstractRange{T}, b::AbstractRange{T}) where T<:Integer
    step(a) == step(b) && return mod(minimum(a) - minimum(b), step(a)) != 0
    invoke(isdisjoint, Tuple{Any,Any}, a, b)
end
```
That implementation seems pretty quick
```julia
julia> @benchmark isdisjoint(x, y) setup=(x=rand(Int):rand(Int); y=rand(Int):rand(Int))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  2.779 ns … 119.561 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     4.159 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   4.380 ns ±   2.237 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

             ▅ ▂ ▃ ▆    ▁       ▇  █   ▂            ▁          
  ▃▆▃▅▄▄▅▂▃▃▅█▄█▄█▆█▅▆▇▃█▅▆▃▃█▄▃█▃▁█▅▂▂█▂▂▅▇▂▂▃▅▂▂▂▃█▂▂▃▂▂▁▂▃ ▄
  2.78 ns         Histogram: frequency by time        6.44 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark isdisjoint(x, y) setup=(x=rand(Int):rand(Int):rand(Int); y=rand(Int):rand(Int):rand(Int))
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):   4.045 ns … 462.406 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):      5.402 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   14.046 ns ±  26.510 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  █▆▂                           ▂▁▁▁                           ▁
  ███▆▆▆▅█▇▅▅▄▁▁▁▃▄▄▁▅▇▅▃▄▄██▇▇▇█████▇▇█▇█▇▆▇▆██▇▇▆▆▅▆▆▃▄▅▅▆▆▆ █
  4.04 ns       Histogram: log(frequency) by time       112 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
But I couldn't benchmark against the existing implementation because `isdisjoint(-6011687643038262928:3545293653953105048, -6446834672754204848:3271267329311042532)` hangs.